proposals/0000-splitting-out-stable-interfaces-from-th.rst

```diff
@@ -0,0 +1,191 @@
+   The code had to be changed to pass an extra ``[]`` argument. See: `the PR to yesod <https://github.com/yesodweb/yesod/pull/1754/files#diff-b0e5dbc5d4ca2998772f987cc5f27c5fc761b34549bdecc93892bbe142d89d26R30>`_.
+
+When upgrading GHC, users are often also forced to upgrade to the new GHC bundled ``template-haskell`` library.
+
+.. _why strategy:
+Justifying our strategy
+^^^^^^^^^^^^^^^^^^^^^^^
+Our strategy is informed by the classes of usages of ``template-haskell`` found in the ecosystem. We can divide users as follows:
+
+* (A) Quote-and-splice clients: These users use only splices, quotes, ``DeriveLift`` or quasiquotes. These users might not even need to import the ``template-haskell`` library.
+* (B) Syntax-construction clients. These users construct Template Haskell syntax trees either directly through its constructors, or indirectly through the smart-constructors exported by ``Language.Haskell.TH.Lib``.
+* (C) Reification clients. These users, notably various forms of deriving, use reification to interrogate the program. Reification currently returns Template Haskell ASTs.
+* (D) Syntax-analysis clients. Some clients pattern match on Template Haskell syntax tree datatypes.
+
+These diverse usages of the library lead to diverse levels of breakage when a new major version of ``template-haskell`` comes out. We can rank them from (A) with the least breakage to (D) with the most.
+For instance, the ``uuid`` library, which just depends on ``template-haskell`` in order to provide a derived ``Lift`` instance (a type (A) client), in all likelyhood would only need to bump its upper-bound on the library.
+On the other-hand ``th-desugar``, which pattern matches on the entire syntax tree (a type (D) client), would have to make code changes on most releases of the library.
+
+Type (A) users are already using interfaces which are quite stable. Yet, they have to update their upper bounds whenever they want to be compatible with a new major version of GHC.
+The first concrete step in our strategy is to publish package that provide these stable APIs. We will return to the benefits of this in the next section.
+
+(B-D) do not currently use stable subsets of the ``template-haskell`` interface.
+In the future, we aim to continue this strategy, by identifying stable interfaces for these classes of users which aren't tightly coupled to the Template Haskell AST.
+The smart-constructors from the ``Language.Haskell.TH.Lib`` module are a good starting point for type (B) clients. Another idea is to use smart-constructors based on the Haskell2010 AST (See: `GHC#20828 <https://gitlab.haskell.org/ghc/ghc/-/issues/20828>`_).
+For type (C) clients, we can build on the existing `th-abstraction` library, and perhaps expose a refined AST that doesn't need to be as expressive as the surface language.
+Type (D) clients on the otherhand are likely to be difficult to accommodate, since they are inherently tightly coupled to the Template Haskell syntax trees.
```

Comment:
```suggestion
Type (D) clients on the other hand are likely to be difficult to accommodate, since they are inherently tightly coupled to the Template Haskell syntax trees.
```